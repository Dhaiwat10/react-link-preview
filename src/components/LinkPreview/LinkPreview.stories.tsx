import React from 'react';
import { storiesOf } from '@storybook/react';
import { LinkPreview } from './LinkPreview';

const customFetcher = async (url: string) => {
  const response = await fetch(`https://rlp-proxy.herokuapp.com/v2?url=${url}`);
  const json = await response.json();
  return json.metadata;
};

storiesOf('LinkPreview', module)
  .add('Default', () => <LinkPreview url='https://barcauniversal.com' />)
  .add('Article', () => (
    <LinkPreview url='https://barcauniversal.com/predicted-barcelona-lineup-against-eibar/' />
  ))
  .add('Text align right', () => (
    <LinkPreview
      url='https://barcauniversal.com/predicted-barcelona-lineup-against-eibar/'
      textAlign='right'
    />
  ))
  .add('Custom image height', () => (
    <LinkPreview
      url='https://barcauniversal.com/predicted-barcelona-lineup-against-eibar/'
      imageHeight='50vh'
    />
  ))
  .add('YouTube link', () => (
    <LinkPreview url='https://www.youtube.com/watch?v=JKJdGNHW1xk' width='30vw' />
  ))
  .add('Twitter link', () => (
    <LinkPreview url='https://twitter.com/BarcaUniversal/status/1396232440314830856' width='20vw' />
  ))
  .add('Reddit link', () => (
    <LinkPreview
      url='https://www.reddit.com/r/LifeProTips/comments/nivqb3/lpt_if_your_your_largest_hex_key_is_to_small_you/'
      width='20vw'
      descriptionLength={100}
    />
  ))
  .add('Fallback', () => <LinkPreview url='https://webzy.dev' fallback={<div>Fallback</div>} />)
  .add('Colors', () => (
    <LinkPreview
      url='https://www.youtube.com/watch?v=dQw4w9WgXcQ'
      backgroundColor='black'
      primaryTextColor='white'
      secondaryTextColor='#ccc'
      borderColor='red'
      width='30vw'
      margin='30px auto'
    />
  ))
  .add('Fallback image', () => (
    <LinkPreview
      url='https://google.com'
      width='30vw'
      fallbackImageSrc='https://www.aljazeera.com/wp-content/uploads/2021/08/2019-12-07T000000Z_879038429_RC2LQD9L67FQ_RTRMADP_3_SOCCER-SPAIN-FCB-RCD-REPORT.jpg?resize=770%2C513'
    />
  ))
  .add('Using custom fetcher', () => (
    <LinkPreview url='stripe.com' fetcher={customFetcher} fallback={<div>Fallback</div>} />
  ))
  .add('Image onError', () => (
    <LinkPreview url='https://www.brianfriel.xyz/learning-how-to-build-on-solana/' width='30vw' />
  ))
  .add('Explicit image', () => (
    <LinkPreview
      url='https://barcauniversal.com/predicted-barcelona-lineup-against-eibar/'
      width='30vw'
      explicitImageSrc='https://barcauniversal.com/wp-content/uploads/2021/05/1002622558-2048x1365.jpg'
    />
  ))
  .add('Explicit no image in case of no image in metadata', () => (
    <LinkPreview url='https://barcauniversal.com' width='30vw' showPlaceholderIfNoImage={false} />
  ));
