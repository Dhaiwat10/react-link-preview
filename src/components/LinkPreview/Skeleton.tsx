import React from 'react';
import LoadingSkeleton from 'react-loading-skeleton';

import './skeleton.scss';
import 'react-loading-skeleton/dist/skeleton.css';

interface SkeletonProps {
  width?: string | number;
  imageHeight?: string | number;
  margin?: string | number;
}

const Skeleton: React.FC<SkeletonProps> = ({ width = '100%', imageHeight = '30vh', margin }) => {
  return (
    <div className='skeleton-container' style={{ width, margin }}>
      <LoadingSkeleton width={width} height={imageHeight} />
      <div className='skeleton-lower-container'>
        <LoadingSkeleton count={3} />
      </div>
    </div>
  );
};

export default Skeleton;
