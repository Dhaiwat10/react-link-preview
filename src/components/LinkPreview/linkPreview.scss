$primary: black;
$secondary: rgb(100, 100, 100);

/* common */
.Container {
  text-align: left;
  background-color: white;
  display: flex;
  flex-direction: column;
  border-radius: 7px;
  border: 1px solid #ccc;
  color: $primary;
  transition: 0.3s all ease;
  height: fit-content;

  &:hover {
    background-color: rgb(250, 250, 250);
    cursor: pointer;
  }
}

.Secondary {
  color: $secondary;
}

.LowerContainer {
  padding: 10px;
}

.Title {
  margin-top: 0;
  margin-bottom: 10px;
}

.Image {
  width: 100%;
  border-top-left-radius: 7px;
  border-top-right-radius: 7px;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
  height: 30vh;
}

.SiteDetails {
  margin-top: 10px;
}

/* sm */
@media (max-width: 640px) {
  .Description {
    display: none;
  }
}

/* md */
@media (min-width: 641px) and (max-width: 768px) {
  .Description {
    display: none;
  }
}
