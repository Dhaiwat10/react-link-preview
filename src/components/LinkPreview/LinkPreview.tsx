import React, { useEffect, useRef, useState } from 'react';

import './linkPreview.scss';
import Skeleton from './Skeleton';

const proxyLink = 'https://rlp-proxy.herokuapp.com/v2?url=';
export const placeholderImg = 'https://i.imgur.com/UeDNBNQ.jpeg';

function isValidResponse(res: APIResponse | null): boolean {
  if (!res) return false;

  return (
    res.title !== null &&
    res.description !== null &&
    res.image !== null &&
    res.siteName !== null &&
    res.hostname !== null &&
    res.title !== undefined &&
    res.description !== undefined &&
    res.image !== undefined &&
    res.siteName !== undefined &&
    res.hostname !== undefined &&
    res.image !== 'null' &&
    !res.image.startsWith('/')
  );
}

export interface LinkPreviewProps {
  url: string;
  className?: string;
  width?: string | number;
  height?: string | number;
  descriptionLength?: number;
  borderRadius?: string | number;
  imageHeight?: string | number;
  textAlign?: 'left' | 'right' | 'center';
  margin?: string | number;
  fallback?: JSX.Element[] | JSX.Element | null;
  backgroundColor?: string;
  primaryTextColor?: string;
  secondaryTextColor?: string;
  borderColor?: string;
  showLoader?: boolean;
  customLoader?: JSX.Element[] | JSX.Element | null;
  openInNewTab?: boolean;
  fetcher?: (url: string) => Promise<APIResponse | null>;
  fallbackImageSrc?: string;
  explicitImageSrc?: string;
  /* Whether the placeholder image is displayed in case no image could be scraped */
  showPlaceholderIfNoImage?: boolean;
}

export interface APIResponse {
  title: string | null;
  description: string | null;
  image: string | null;
  siteName: string | null;
  hostname: string | null;
}

export const LinkPreview: React.FC<LinkPreviewProps> = ({
  url,
  className = '',
  width,
  height,
  descriptionLength,
  borderRadius,
  imageHeight,
  textAlign,
  margin,
  fallback = null,
  backgroundColor = 'white',
  primaryTextColor = 'black',
  secondaryTextColor = 'rgb(100, 100, 100)',
  borderColor = '#ccc',
  showLoader = true,
  customLoader = null,
  openInNewTab = true,
  fetcher,
  fallbackImageSrc = placeholderImg,
  explicitImageSrc = null,
  showPlaceholderIfNoImage = true,
}) => {
  const _isMounted = useRef(true);
  const [metadata, setMetadata] = useState<APIResponse | null>();
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    _isMounted.current = true;
    setLoading(true);

    if (fetcher) {
      fetcher(url)
        .then((res) => {
          if (_isMounted.current) {
            if (isValidResponse(res)) {
              setMetadata(res);
            } else {
              setMetadata(null);
            }
            setLoading(false);
          }
        })
        .catch((err: Error) => {
          console.error(err);
          console.error('No metadata could be found for the given URL.');
          if (_isMounted.current) {
            setMetadata(null);
            setLoading(false);
          }
        });
    } else {
      fetch(proxyLink + url)
        .then((res) => res.json())
        .then((res) => {
          if (_isMounted.current) {
            setMetadata((res.metadata as unknown) as APIResponse);
            setLoading(false);
          }
        })
        .catch((err: Error) => {
          console.error(err);
          console.error('No metadata could be found for the given URL.');
          if (_isMounted.current) {
            setMetadata(null);
            setLoading(false);
          }
        });
    }

    return () => {
      _isMounted.current = false;
    };
  }, [url, fetcher]);

  if (loading && showLoader) {
    if (customLoader) {
      return <>{customLoader}</>;
    } else {
      return <Skeleton width={width} imageHeight={imageHeight} margin={margin} />;
    }
  }

  if (!metadata) {
    return <>{fallback}</>;
  }

  const { image, description, title, siteName, hostname } = metadata;

  const onClick = () => {
    const browserTarget = openInNewTab ? '_blank' : '_self';
    window.open(url, browserTarget);
  };

  return (
    <div
      data-testid='container'
      onClick={onClick}
      className={`Container ${className}`}
      style={{ width, height, borderRadius, textAlign, margin, backgroundColor, borderColor }}
    >
      {(image || fallbackImageSrc) && showPlaceholderIfNoImage && (
        <div
          data-testid='image-container'
          style={{
            borderTopLeftRadius: borderRadius,
            borderTopRightRadius: borderRadius,
            backgroundImage: `url(${
              explicitImageSrc || image || fallbackImageSrc
            }), url(${fallbackImageSrc})`,
            height: imageHeight,
          }}
          className='Image'
        ></div>
      )}

      <div className='LowerContainer'>
        <h3 data-testid='title' className='Title' style={{ color: primaryTextColor }}>
          {title}
        </h3>
        {description && (
          <span
            data-testid='desc'
            className='Description Secondary'
            style={{ color: secondaryTextColor }}
          >
            {descriptionLength
              ? description.length > descriptionLength
                ? description.slice(0, descriptionLength) + '...'
                : description
              : description}
          </span>
        )}
        <div className='Secondary SiteDetails' style={{ color: secondaryTextColor }}>
          {siteName && <span>{siteName} â€¢ </span>}
          <span>{hostname}</span>
        </div>
      </div>
    </div>
  );
};
