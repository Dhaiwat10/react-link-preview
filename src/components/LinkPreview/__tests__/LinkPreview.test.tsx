import React from 'react';
import { LinkPreview } from '..';
import { render, screen } from '@testing-library/react';
import fetchMock from 'fetch-mock';
import { APIResponse } from '../LinkPreview';

const url = 'https://barcauniversal.com/predicted-barcelona-lineup-against-eibar/';

const metadata: APIResponse = {
  title: 'Predicted Barcelona lineup against Eibar | Barca Universal',
  description:
    'For one final time this season, possibly for one last time under Ronald Koeman, Barcelona will take the field on Saturday. The season took a nosedive that few foresaw but left none surprised. Though Barcelona plays the relegated Eibar at the Ipurua, all eyes will be on the title decisive clashes of the two Madrid […]',
  image: 'https://barcauniversal.com/wp-content/uploads/2021/05/1002622558-scaled.jpg',
  siteName: 'Barca Universal',
  hostname: 'barcauniversal.com',
};

const customFetcher = async (url: string) => {
  const response = await fetch(`https://rlp-proxy.herokuapp.com/v2?url=${url}`);
  const json = await response.json();
  return json.metadata;
};

fetchMock.get(`https://rlp-proxy.herokuapp.com/v2?url=${url}`, {
  metadata,
});

describe('LinkPreview Component', () => {
  it('renders the component correctly', async () => {
    const { container } = render(<LinkPreview url={url} />);
    expect(container).toMatchSnapshot();
    const title = await screen.findByText(metadata.title as string);
    const description = await screen.findByText(metadata.description as string);
    const siteName = await screen.findByText(`${metadata.siteName} •`);
    const hostname = await screen.findByText(metadata.hostname as string);
    expect(title).toBeTruthy();
    expect(description).toBeTruthy();
    expect(siteName).toBeTruthy();
    expect(hostname).toBeTruthy();
  });

  it('applies the className prop', async () => {
    const className = 'TestClass';

    render(<LinkPreview url={url} className={className} />);

    const containerDiv = await screen.findByTestId('container');
    expect(containerDiv.classList.contains(className)).toBeTruthy;
  });

  it('applies the various style props correctly', async () => {
    const width = '400px';
    const height = '400px';
    const borderRadius = '10px';
    const imageHeight = '100px';
    const textAlign = 'left';
    const margin = '10px';
    const backgroundColor = 'rgb(204, 204, 204)';
    const primaryTextColor = 'rgb(255, 255, 255)';
    const secondaryTextColor = 'rgb(240, 240, 240)';
    const borderColor = 'rgb(0, 0, 0)';

    render(
      <LinkPreview
        url={url}
        width={width}
        height={height}
        borderRadius={borderRadius}
        imageHeight={imageHeight}
        textAlign={textAlign}
        margin={margin}
        primaryTextColor={primaryTextColor}
        secondaryTextColor={secondaryTextColor}
        backgroundColor={backgroundColor}
        borderColor={borderColor}
      />
    );

    const containerDiv = await screen.findByTestId('container');
    expect(containerDiv).toMatchSnapshot();
    expect(containerDiv.style.width).toBe(width);
    expect(containerDiv.style.height).toBe(height);
    expect(containerDiv.style.textAlign).toBe(textAlign);
    expect(containerDiv.style.margin).toBe(margin);
    expect(containerDiv.style.backgroundColor).toBe(backgroundColor);
    expect(containerDiv.style.borderColor).toBe(borderColor);

    const innerContainer = await screen.findByTestId('image-container');
    expect(innerContainer.style.borderTopLeftRadius).toBe(borderRadius);
    expect(innerContainer.style.borderTopRightRadius).toBe(borderRadius);
    expect(innerContainer.style.height).toBe(imageHeight);

    const title = await screen.findByTestId('title');
    expect(title.style.color).toBe(primaryTextColor);

    const desc = await screen.findByTestId('desc');
    expect(desc.style.color).toBe(secondaryTextColor);
  });

  it('renders fallback correctly', async () => {
    const url = 'https://google.com';
    const fallback = <div data-testid='fallback'>fallback</div>;

    fetchMock.getOnce(`https://rlp-proxy.herokuapp.com/v2?url=${url}`, {
      metadata: null,
    });

    const { container } = render(<LinkPreview url={url} fallback={fallback} />);
    expect(container).toMatchSnapshot();
    const fallbackDiv = await screen.findByTestId('fallback');
    expect(fallbackDiv).toBeTruthy();
  });

  it('works properly with a custom fetcher', async () => {
    render(<LinkPreview url={url} fetcher={customFetcher} />);

    const title = await screen.findByText(metadata.title as string);
    const description = await screen.findByText(metadata.description as string);
    const siteName = await screen.findByText(`${metadata.siteName} •`);
    const hostname = await screen.findByText(metadata.hostname as string);
    // const containerDiv = await screen.findByTestId('image-container');
    expect(title).toBeTruthy();
    expect(description).toBeTruthy();
    expect(siteName).toBeTruthy();
    expect(hostname).toBeTruthy();
    // expect(containerDiv.style.backgroundImage).toBe(
    //   `url(${metadata.image}), url(${placeholderImg})`
    // );
  });

  it('calls onSuccess after fetching metadata', async () => {
    const onSuccess = jest.fn((metadata) => console.log(metadata));

    render(<LinkPreview url={url} onSuccess={onSuccess} />);

    setTimeout(() => {
      expect(onSuccess).toBeCalledTimes(1);
    }, 1000);
  });
});
