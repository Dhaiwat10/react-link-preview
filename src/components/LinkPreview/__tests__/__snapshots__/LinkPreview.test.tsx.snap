// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkPreview Component applies the various style props correctly 1`] = `
<div
  class="Container "
  data-testid="container"
  style="width: 400px; height: 400px; border-radius: 10px; text-align: left; margin: 10px; background-color: rgb(204, 204, 204); border-color: rgb(0, 0, 0);"
>
  <div
    class="Image"
    data-testid="image-container"
    style="border-top-left-radius: 10px; border-top-right-radius: 10px; height: 100px;"
  />
  <div
    class="LowerContainer"
  >
    <h3
      class="Title"
      data-testid="title"
      style="color: rgb(255, 255, 255);"
    >
      Predicted Barcelona lineup against Eibar | Barca Universal
    </h3>
    <span
      class="Description Secondary"
      data-testid="desc"
      style="color: rgb(240, 240, 240);"
    >
      For one final time this season, possibly for one last time under Ronald Koeman, Barcelona will take the field on Saturday. The season took a nosedive that few foresaw but left none surprised. Though Barcelona plays the relegated Eibar at the Ipurua, all eyes will be on the title decisive clashes of the two Madrid […]
    </span>
    <div
      class="Secondary SiteDetails"
      style="color: rgb(240, 240, 240);"
    >
      <span>
        Barca Universal
         • 
      </span>
      <span>
        barcauniversal.com
      </span>
    </div>
  </div>
</div>
`;

exports[`LinkPreview Component renders fallback correctly 1`] = `
<div>
  <div
    class="skeleton-container"
    style="width: 100%;"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="width: 100%; height: 30vh;"
      >
        ‌
      </span>
      <br />
    </span>
    <div
      class="skeleton-lower-container"
    >
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
        >
          ‌
        </span>
        <br />
        <span
          class="react-loading-skeleton"
        >
          ‌
        </span>
        <br />
        <span
          class="react-loading-skeleton"
        >
          ‌
        </span>
        <br />
      </span>
    </div>
  </div>
</div>
`;

exports[`LinkPreview Component renders the component correctly 1`] = `
<div>
  <div
    class="skeleton-container"
    style="width: 100%;"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="width: 100%; height: 30vh;"
      >
        ‌
      </span>
      <br />
    </span>
    <div
      class="skeleton-lower-container"
    >
      <span
        aria-busy="true"
        aria-live="polite"
      >
        <span
          class="react-loading-skeleton"
        >
          ‌
        </span>
        <br />
        <span
          class="react-loading-skeleton"
        >
          ‌
        </span>
        <br />
        <span
          class="react-loading-skeleton"
        >
          ‌
        </span>
        <br />
      </span>
    </div>
  </div>
</div>
`;
